### ISSUE TITLE
Exclusion based on list provided in SQL query string

### ISSUE THREAD

**Issue by jrobiii** [olahallengren/sql-server-maintenance-solution#95]
I have an issue running OLA Jobs via CMS (MSX/TSX) and excluding specific databases for specific instances.  Where as, if we change the job it will affect the job on all servers that subscribe to that job. 

I propose a new parameter (in my case for DatabaseIngegrityCheck, but could apply to others) that would take a query string.  The results of the query string would be a list of databases to remove from the ALL_DATABASES collection.  I would invision something like:
`
execute DBAS.dbo.DatabaseIntegrityCheck
    @Databases = 'ALL_Databases', 
    @ExclusionQuery = 'select DatabaseName from DBAS.dbo.OlaExclusions where ProcedureName = ''DatabaseIntegrityCheck'''
`

You could have any query you like as long as it returned a list of DatabaseNames.  Say that you want to exclude system databases
`
execute DBAS.dbo.DatabaseIntegrityCheck
    @Databases = 'ALL_Databases', 
    @ExclusionQuery = 'select name as DatabaseName from sys.databases where dbid <=4'
`

Thanks,

**Comment by drstonephd** [comment]
I think this is only a partial solution.  It sounds like neither of us want to define our MSX jobs one per instance, so we need a way to customize parameters that's identical on all servers.  Because the other procedure parameters should be configurable, perhaps a local config table of some sort that stores the parameters?  However, this defeats the central control of the jobs.  We could define a master job to configure the values in the local tables based on server and instance name...or include the server-instance name in the config table.

I also want to configure job properties such as notification and schedules.  The job settings can't be done nicely with MSX/TSX.  I was thinking of using MSX/TSX to deploy a job that installs/maintains the maintenance version and defines/controls local jobs.

**Comment by drstonephd** [comment]
Another option would be to define a procedure the returns the parameters to feed the maintenance procedure.  The procedure could test the server-instance name and return the values desired for that instance.  The configuration would be in one procedure for all instances.  This would allow the same job to behave differently on different instances.   Also, the maintenance procedures would not need to be updated...just the master jobs that call them.

**Comment by drstonephd** [comment]
Add a job step - first - that insures the local job properties like notifications and schedules are those desired for that instance, based on a config procedure or table?

**Comment by jrobiii** [comment]
Thank you drstonephd for your suggestions.

I guess I wasn't clear on my objective.  I need to provide a dynamic database exclusion list for "ALL_Databases".

With that in mind, what I suggest is a complete solution to meet this objective.

**Comment by jrobiii** [comment]
Sorry, clicked on comment before I was finished.  

Additionally, with my suggestion, I only need one MX job to DatabaseIntegrityCheck on all servers. In my case the exclusions are maintained on the target server, but with my suggestion you could have any query you want. So if you want to call a store procedure/function you can still do that.

**Comment by drstonephd** [comment]
Yes, that would work.  I have been thinking about this for a while and have been trying to avoid having a solution that requires defining a local configuration locally.  I would like a single master job or script that handles custom local procedure parameters and job properties (e.g., schedules).  I don't think writing a query for each parameter would be general solution.  I would rather have a local table that has configuration data for all maintenance jobs - procedure parameters and job schedules.  If you need dynamic setting, then perhaps a procedure to update the configuration table as required?  

To avoid defining the local configuration locally, I was thinking about writing a PowerShell job to sync a configuration table from the master to all targets.  A PowerShell job could sync back to the master server the instance ComandLog data to a "MasterComandLog" table.  

This would have to be an add-on that does not require changes to Ola's script.  However, if Ola adds a configuration table for procedure parameters and job properties, then it would be a lot easier to setup a master.  It would also improve the single instance setup experience in that if we define the configuration before running Ola's scripts, his script could set the parameters and job properties, including schedules.  His current defaults would still be used (e.g., no schedules), if there is no configuration data.  

When a new version comes out, I sometimes wonder if I should recreate the jobs.  If the script used a configuration table, then I could drop the jobs and have the setup script use my config data to recreate the jobs.

**Comment by gnotisauton** [comment]
Basic question @jrobiii : why are you excluding databases from checkDB?

**Comment by jrobiii** [comment]
We have a few very large static databases. DBCC CHECKDB will not complete
during our weekly maintenance window. This would be three or four databases
out thousands.
